### PR TITLE
Feature/guest check in

### PR DESCRIPTION
**Pull request by zigavehovec**
This PR adds guest check-in/check-out functionality:
image

image

A couple UI improvements were also added:
- In my opinion more appropriate icons + minor sidebar re-organization
image

- Toolbar view, which removes a lot of duplicate code + adds back button support
image

***Please double check the new strings, thanks***

### CODE REVIEW

File: src/moderatorFrontendMain/kotlin/app/AppContent.kt [hunk 1]
```diff
@@ -101,12 +65,6 @@ interface AppContentClasses {
 private val AppContentStyle = { theme: dynamic ->
   // Keep in sync with AppContentClasses!
   js {
```
Comment (vonox7) [inline src/moderatorFrontendMain/kotlin/app/AppContent.kt:67]: container is now undefined? should we keep it in sync as written in the comment?

File: src/moderatorFrontendMain/kotlin/util/Strings.kt [hunk 2]
```diff
@@ -36,6 +36,16 @@ object Strings {
...
   )
 
+  val other = LocalizedString (
+    "Other",
+    "Sonstige",
+  )
+
+  val invalid_email = LocalizedString(
+    "Invalid email address",
+    "Ungültige E-Mail-Adresse",
```
Comment (vonox7) [inline src/moderatorFrontendMain/kotlin/util/Strings.kt:46]:
```suggestion
    "Ungültige E-Mail Adresse",
```

File: src/moderatorFrontendMain/kotlin/util/Strings.kt [hunk 3]
```diff
@@ -610,6 +625,51 @@ object Strings {
...
   )
 
+  val guest_checkin = LocalizedString(
+    "Guest check in",
+    "Gast Check-In",
+  )
+
+  val guest_checkin_add_guest = LocalizedString(
+    "Add guest",
+    "Gast hinzufügen",
```
Comment (vonox7) [inline src/moderatorFrontendMain/kotlin/util/Strings.kt:635]:
```suggestion
    "Check in guest",
    "Gast einchecken",
```

File: src/moderatorFrontendMain/kotlin/util/Strings.kt [hunk 4]
```diff
@@ -610,6 +625,51 @@ object Strings {
...
+  )
+
+  val guest_checkin_not_yet_added_title = LocalizedString(
+    "No guest check in added yet.",
+    "Noch keine Gast Check-In hinzugefügt.",
```
Comment (vonox7) [inline src/moderatorFrontendMain/kotlin/util/Strings.kt:640]:
```suggestion
    "Currently no guests are checked in.",
    "Momentan sind keine Gäste eingecheckt.",
```

File: src/moderatorFrontendMain/kotlin/util/Strings.kt [hunk 5]
```diff
@@ -610,6 +625,51 @@ object Strings {
...
+  )
+
+  val guest_checkin_not_yet_added_subtitle = LocalizedString(
+    "Click on \"add guest\" button in the top right corner to add a guest check in.",
+    "Klicken Sie rechts oben auf \"Gast hinzufügen\" um eine Gast Check-In hinzuzufügen.",
```
Comment (vonox7) [inline src/moderatorFrontendMain/kotlin/util/Strings.kt:645]:
```suggestion
    "Click on \"check in guest\" in the top right corner to check in a guest.",
    "Klicken Sie rechts oben auf \"Gast einchecken\" um einen Gast einzuchecken.",
```

File: src/moderatorFrontendMain/kotlin/util/Strings.kt [hunk 6]
```diff
@@ -610,6 +625,51 @@ object Strings {
...
+  )
+
+  val guest_checkin_checkout_are_you_sure = LocalizedString(
+    "Are you sure you want to check out %s?",
+    "Sind Sie sicher, dass Sie %s auschecken möchten?", // FIXME: auschecken
```
Comment (vonox7) [inline src/moderatorFrontendMain/kotlin/util/Strings.kt:650]:
```suggestion
    "Check out %s now?",
    "%s jetzt auschecken?",
```

File: src/moderatorFrontendMain/kotlin/util/Strings.kt [hunk 7]
```diff
@@ -610,6 +625,51 @@ object Strings {
...
+  )
+
+  val guest_checkin_email_must_not_be_empty = LocalizedString(
+    "Please add an email address",
+    "Bitte fügen Sie eine E-Mail-Adresse hinzu",
```
Comment (vonox7) [inline src/moderatorFrontendMain/kotlin/util/Strings.kt:655]:
```suggestion
    "Please add the email address of the guest",
    "Bitte fügen Sie die E-Mail Adresse des Gasts hinzu",
```

File: src/moderatorFrontendMain/kotlin/util/Urls.kt [hunk 8]
```diff
@@ -17,6 +17,7 @@ enum class Url(val path: String, val title: LocalizedString, val requiresAuth: B
     Strings.access_control,
     requiresAuth = true
   ),
+  GUEST_CHECK_IN("$pathBase/guest-access", Strings.guest_checkin, requiresAuth = true),
```
Comment (vonox7) [inline src/moderatorFrontendMain/kotlin/util/Urls.kt:20]:
```suggestion
  GUEST_CHECK_IN("$pathBase/check-in", Strings.guest_checkin, requiresAuth = true),
```

File: src/serverMain/kotlin/com/studo/campusqr/Server.kt [hunk 9]
```diff
@@ -103,6 +103,7 @@ suspend fun main() {
 
         route("{id}") {
           post("visit") { call.visitLocation() }
+          post("guestVisit") { call.getAuthenticatedCall()?.guestVisitLocation() }
```
Comment (vonox7) [inline src/serverMain/kotlin/com/studo/campusqr/Server.kt:106]: is "guestVisit" meant to be "guestCheckIn"?

File: src/serverMain/kotlin/com/studo/campusqr/database/MainDatabase.kt [hunk 10]
```diff
@@ -33,6 +33,9 @@ object MainDatabase : MongoDatabase(mongoUri, collections = {
             CheckIn::checkOutDate.descending(), CheckIn::date.descending(),
             partialIndex = arrayOf(CheckIn::checkOutDate equal null)
         )
+
+        // Guest check in / check out
+        index(CheckIn::checkedInBy.ascending(), CheckIn::checkOutDate.descending())
```
Comment (vonox7) [inline src/serverMain/kotlin/com/studo/campusqr/database/MainDatabase.kt:38]: partialIndex: `checkedInBy notEqual null, checkOutDate equal null`

File: src/serverMain/kotlin/com/studo/campusqr/database/Payloads.kt [hunk 11]
```diff
@@ -101,6 +101,7 @@ class CheckIn : MongoMainEntry() {
   var ipAddress: String? = null // Only stored if checkInIpAddressHeader is set
   var grantAccessId: String? = null // id of BackendAccess which was used to enter, null if no BackendAccess was used
   var seat: Int? = null // null if Location has no seatCount defined
+  var checkedInBy: String? = null // userId of the person who managed this a manual guest check-in
```
Comment (vonox7) [inline src/serverMain/kotlin/com/studo/campusqr/database/Payloads.kt:104]: I don't understand this comment

File: src/serverMain/kotlin/com/studo/campusqr/endpoints/LocationManagement.kt [hunk 12]
```diff
@@ -161,6 +163,14 @@ suspend fun ApplicationCall.visitLocation() {
   respondOk()
 }
 
+suspend fun AuthenticatedApplicationCall.guestVisitLocation() {
+  if (user.type == UserType.ACCESS_MANAGER || user.isAdmin) {
```
Comment (vonox7) [inline src/serverMain/kotlin/com/studo/campusqr/endpoints/LocationManagement.kt:167]: what about moderators? when you ask for user.isAccessManager then i think we don't have this problem

File: src/serverMain/kotlin/com/studo/campusqr/endpoints/ReportData.kt [hunk 13]
```diff
@@ -251,21 +249,53 @@ suspend fun AuthenticatedApplicationCall.listAllActiveCheckIns() {
...
+        seat = checkIn.seat,
+        checkInDate = checkIn.date.time.toDouble(),
+        email = checkIn.email
+      )
+    }
+  )
+}
+
+suspend fun AuthenticatedApplicationCall.listGuestActiveCheckIns() {
+  if (user.type != UserType.ACCESS_MANAGER && !user.isAdmin) {
```
Comment (vonox7) [inline src/serverMain/kotlin/com/studo/campusqr/endpoints/ReportData.kt:269]: here too

File: src/serverMain/kotlin/com/studo/campusqr/endpoints/ReportData.kt [hunk 14]
```diff
@@ -251,21 +249,53 @@ suspend fun AuthenticatedApplicationCall.listAllActiveCheckIns() {
...
         seat = checkIn.seat,
-        checkInDate = checkIn.date.time.toDouble()
+        checkInDate = checkIn.date.time.toDouble(),
+        email = checkIn.email
       )
     }
   )
 }
+
+suspend fun List<CheckIn>.getLocationMap(): Map<String, BackendLocation> = runOnDb {
```
Comment (vonox7) [inline src/serverMain/kotlin/com/studo/campusqr/endpoints/ReportData.kt:297]: Pls comment the return type: String is not clear what this is.

**Review by vonox7 — CHANGES_REQUESTED** [review]

**Comment by vonox7** [comment]
@functionaldude pls check the backend side of this PR

File: src/serverMain/kotlin/com/studo/campusqr/Server.kt (hunk 9 above)
Comment (functionaldude) [reply src/serverMain/kotlin/com/studo/campusqr/Server.kt:106]: if yes, this should probably not use the `getAuthenticatedCall()`

**Review by functionaldude — APPROVED** [review]

**Review by vonox7 — APPROVED** [review]

**Pull request merged**